pi5/pinctrl_utils.go

```diff
@@ -0,0 +1,287 @@
+	until we get the child's physical address in all of /dev/gpiomem. */
+	b.physAddr, err = setGPIONodePhysAddrHelper(currNodePath, INVALID_ADDR, numCAddrCells)
+	if err != nil {
+		return fmt.Errorf("trouble calculating phys addr for %s: %w\n", nodePath, err)
+	}
+
+	return nil
+}
+
+// Recursively Traverses Device Tree to Calcuate Physical Address of specified GPIO Chip
+func setGPIONodePhysAddrHelper(currNodePath string, physAddress uint64, numCAddrCells uint32) (uint64, error) {
+
+	if currNodePath == dtBaseNodePath { // Base Case: We are at the root of the device tree.
+		return physAddress, nil
+	}
+
+	// Normal Case: We are not at the root of the device tree. We must continue mapping our child addr (from the previous call) to this parent's addr space.
+	parentNodePath := filepath.Dir(currNodePath)
+	numPAddrCells, numAddrSpaceCells, err := getNumAddrSizeCellsInfo(parentNodePath)
+	if err != nil {
+		return INVALID_ADDR, err
+	}
+
+	var addrRanges []rangeInfo
+	if physAddress == INVALID_ADDR { // Case 1: We are the Child Node. No addr has been set. Read the reg file to get the physical address within our parents space.
+		physAddress, err = getRegAddr(currNodePath, numPAddrCells)
+		if err != nil {
+			return INVALID_ADDR, err
+		}
+
+	} else { // Case 2: We use the ranges property to continue mapping a child addr into our parent addr space.
+		addrRanges, err = getRangesAddrInfo(currNodePath, numCAddrCells, numPAddrCells, numAddrSpaceCells)
+		if err != nil {
+			return INVALID_ADDR, err
+		}
+
+		// getRangesAddrInfo returns a list of all possible child address ranges our physical address can fall into. We must see which range to use, so that we can map our physical address into the correct parent address range.
+		for _, addrRange := range addrRanges {
+
+			if addrRange.childAddr <= physAddress && physAddress <= addrRange.childAddr+addrRange.addrSpaceSize {
+				physAddress -= addrRange.childAddr  // get the offset beween the address and child base address
+				physAddress += addrRange.parentAddr // now address has been mapped into parent space.
+				break
+			}
+		}
+	}
+
+	numCAddrCells = numPAddrCells
+	currNodePath = parentNodePath
+	return setGPIONodePhysAddrHelper(currNodePath, physAddress, numCAddrCells)
+}
+
+// Creates a virtual Page to access/manipulate memory related to gpiochip data
+func (b *pinctrlpi5) createGPIOVPage(memPath string) error {
+	var err error
+
+	// Open the 'file' you are trying to map. Note: /dev/gpiomem0 is an alias, so it's file length returns 0
+	fileFlags := os.O_RDWR | os.O_SYNC
+	b.memFile, err = os.OpenFile("/dev/gpiomem0", fileFlags, 0666) // 0666 is an octal representation of: file is readable / writeable by anyone
+	if err != nil {
+		return fmt.Errorf("failed to open %s: %w\n", memPath, err)
+	}
+
+	pageSize := uint64(syscall.Getpagesize())
+	pageOffset := b.physAddr & (pageSize - 1) // difference between base address of the page and the address we're actually tring to access
```

Comment:
optional change but I prefer comments above the code theyre referencing - makes it more readable